src/ORM/RelatedData/StandardRelatedDataService.php

```diff
@@ -0,0 +1,507 @@
+     *
+     * @param array $classIDs
+     * @param DataObject $record
+     * @param string[] $throughClasses
+     */
+    private function addRelatedManyManys(array &$classIDs, DataObject $record, array &$throughClasses): void
+    {
+        $class = get_class($record);
+        foreach ($record->manyMany() as $component => $componentClass) {
+            $componentClass = $this->updateComponentClass($componentClass, $throughClasses);
+            if (
+                // Ignore belongs_many_many_through with dot syntax
+                strpos($componentClass, '.') !== false ||
+                // Prevent duplicate counting of self-referential relations e.g.
+                // MyFile::$many_many = [ 'MyFile' => MyFile::class ]
+                // This relation will still be counted in $this::addRelatedReverseManyManys()
+                $record instanceof $componentClass
+            ) {
+                continue;
+            }
+            $results = $this->fetchManyManyResults($record, $class, $component, false);
+            $this->addResultsToClassIDs($classIDs, $results, $componentClass);
+        }
+    }
+
+    /**
+     * Query the database to retrieve many-many results
+     *
+     * @param DataObject $record - The DataObject whose usage data is being retrieved, usually a File
+     * @param string $class - example: My\App\SomePageType
+     * @param string $component - example: 'SomeFiles' - My\App\SomePageType::SomeFiles()
+     * @param bool $reverse - true: SomePage::SomeFiles(), false: SomeFile::SomePages()
+     * @return Query|null
+     */
+    private function fetchManyManyResults(
+        DataObject $record,
+        string $class,
+        string $component,
+        bool $reverse
+    ): ?Query {
+        // Example php file: class MyPage ... private static $many_many = [ 'MyFile' => File::class ]
+        $data = $this->dataObjectSchema->manyManyComponent($class, $component);
+        if (!$data || !($data['join'] ?? false)) {
+            return null;
+        }
+        $joinTableName = $this->deriveJoinTableName($data);
+        if (!ClassInfo::hasTable($joinTableName)) {
+            return null;
+        }
+        $usesThroughTable = $data['join'] != $joinTableName;
+
+        $parentField = preg_replace('#ID$#', '', $data['parentField']) . 'ID';
+        $childField = preg_replace('#ID$#', '', $data['childField']) . 'ID';
+        $selectField = !$reverse ? $childField : $parentField;
+        $selectFields = [$selectField];
+        $whereField = !$reverse ? $parentField : $childField;
+
+        // Support for polymorphic through objects such FileLink that allow for multiple class types on one side e.g.
+        // ParentID: int, ParentClass: enum('File::class, SiteTree::class, ElementContent::class, ...')
+        if ($usesThroughTable) {
+            $dbFields = $this->dataObjectSchema->databaseFields($data['join']);
+            if ($parentField === 'ParentID' && isset($dbFields['ParentClass'])) {
```

Comment:
Again, assuming that ParentID is specifically the only polymorphic field? Same critique as above.